File: src_dev/reinforcement_learning/rl_dqn.py

import torch
import torch.nn as nn
import torch.optim as optim
import numpy as np
import random
from collections import deque
import time

# Check description: "Deep value iteration methods align with the objective... 
# This includes Q-learning [161] and its deep-learning version—DQN [162]" (Lines 104, 122)

# Set device
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f"Using device: {device}")

# ==========================================
# PART 1: CONTINUOUS ENVIRONMENT
# ==========================================
# Making this "Practically Relevant" by using Continuous State Space (Position, Velocity)
# Goal: Stabilize a particle at Position = 0.
class ContinuousWorld:
    def __init__(self):
        # State: [Position, Velocity] (Continuous)
        self.state_size = 2
        # Action: 0 (Push Left), 1 (Do Nothing), 2 (Push Right) (Discrete)
        self.action_size = 3
        
        self.state = np.array([0.0, 0.0])
        self.max_steps = 200
        self.current_step = 0
        
    def reset(self):
        # Start at random position away from center
        pos = random.uniform(-2.0, 2.0)
        vel = 0.0
        self.state = np.array([pos, vel])
        self.current_step = 0
        return self.state

    def step(self, action):
        pos, vel = self.state
        
        # Physics Dynamics
        force = 0.0
        if action == 0: force = -1.0 # Left
        elif action == 2: force = 1.0 # Right
        
        vel += force * 0.1 # F=ma (dt=0.1)
        pos += vel * 0.1
        
        # Clip to keeping boundaries reasonable
        pos = np.clip(pos, -5.0, 5.0)
        vel = np.clip(vel, -2.0, 2.0)
        
        self.state = np.array([pos, vel])
        self.current_step += 1
        
        # Reward Function
        # We want: Distance to be 0
        reward = -np.abs(pos) - 0.1 * np.abs(vel) # Penalty for distance and speed (Energy min)
        
        # Done condition
        done = self.current_step >= self.max_steps
        
        return self.state, reward, done

# ==========================================
# PART 2: THE NETWORK (Q-Function)
# ==========================================
# Ref: "In policy gradient methods... theta can be the weights and biases of a deep neural network" (Lines 88-89)
# Here we use it for Value Function (DQN)
class QNetwork(nn.Module):
    def __init__(self, state_size, action_size):
        super(QNetwork, self).__init__()
        # Simple feed-forward network
        self.fc1 = nn.Linear(state_size, 64)
        self.fc2 = nn.Linear(64, 64)
        self.fc3 = nn.Linear(64, action_size) # Outputs Q-value for each action
        
    def forward(self, x):
        # x is already a tensor on the correct device if passed correctly, 
        # but for safety in the policy method we usually cast it.
        if not isinstance(x, torch.Tensor):
            x = torch.tensor(x, dtype=torch.float32).to(device)
        return self.fc3(torch.relu(self.fc2(torch.relu(self.fc1(x)))))

# ==========================================
# PART 3: REPLAY BUFFER
# ==========================================
# Essential for Deep RL to break correlations in training data
class ReplayBuffer:
    def __init__(self, capacity=10000):
        self.buffer = deque(maxlen=capacity)
    
    def push(self, state, action, reward, next_state, done):
        self.buffer.append((state, action, reward, next_state, done))
        
    def sample(self, batch_size):
        return random.sample(self.buffer, batch_size)
    
    def __len__(self):
        return len(self.buffer)

# ==========================================
# PART 4: DQN AGENT
# ==========================================
class DQNAgent:
    def __init__(self, state_size, action_size):
        self.state_size = state_size
        self.action_size = action_size
        
        # Hyperparameters
        self.gamma = 0.99
        self.epsilon = 1.0
        self.epsilon_min = 0.05
        self.epsilon_decay = 0.995
        self.batch_size = 64
        self.lr = 0.001
        
        # Networks
        self.q_net = QNetwork(state_size, action_size).to(device)
            
        self.optimizer = optim.Adam(self.q_net.parameters(), lr=self.lr)
        self.memory = ReplayBuffer()

    def policy(self, state):
        # Epsilon-Greedy
        if random.random() < self.epsilon:
            return random.randint(0, self.action_size - 1)
        else:
            with torch.no_grad():
                q_values = self.q_net(state)
                return torch.argmax(q_values).item()

    def learn(self):
        if len(self.memory) < self.batch_size:
            return
        
        # Sample batch
        batch = self.memory.sample(self.batch_size)
        states, actions, rewards, next_states, dones = zip(*batch)
        
        # Convert to tensors and move to device
        states = torch.tensor(np.array(states), dtype=torch.float32).to(device)
        actions = torch.tensor(actions, dtype=torch.long).unsqueeze(1).to(device) # Add dim for gather
        rewards = torch.tensor(rewards, dtype=torch.float32).unsqueeze(1).to(device)
        next_states = torch.tensor(np.array(next_states), dtype=torch.float32).to(device)
        dones = torch.tensor(dones, dtype=torch.float32).unsqueeze(1).to(device)

        # Compute Q(s, a)
        # self.q_net(states) gives [Batch, ActionSize]
        # .gather(1, actions) selects the Q-value for the specific action taken
        current_q = self.q_net(states).gather(1, actions)
        
        # Compute Target: r + gamma * max Q(s')
        # We use .max(1)[0] to get max along action dim
        with torch.no_grad():
            next_q = self.q_net(next_states).max(1)[0].unsqueeze(1)
            target_q = rewards + (1 - dones) * self.gamma * next_q
            
        # Loss (MSE)
        # "maximize the cumulative expected reward G" (Line 36) -> Minimize prediction error
        loss = nn.MSELoss()(current_q, target_q)
        
        # Optimize
        self.optimizer.zero_grad()
        loss.backward()
        self.optimizer.step()
        
    def decay_epsilon(self):
        if self.epsilon > self.epsilon_min:
            self.epsilon *= self.epsilon_decay

# ==========================================
# PART 5: TRAINING LOOP
# ==========================================
def run_dqn():
    env = ContinuousWorld()
    agent = DQNAgent(env.state_size, env.action_size)
    
    num_episodes = 1000
    
    print("Starting DQN Training (PyTorch)...")
    print("Goal: Stabilize particle at 0.0.")
    
    for episode in range(num_episodes):
        state = env.reset()
        total_reward = 0
        done = False
        
        while not done:
            action = agent.policy(state)
            next_state, reward, done = env.step(action)
            
            # Store experience
            agent.memory.push(state, action, reward, next_state, done)
            
            # Learn from batch
            agent.learn()
            
            state = next_state
            total_reward += reward
            
        agent.decay_epsilon()
        
        if (episode + 1) % 10 == 0:
            print(f"Episode {episode+1}/{num_episodes} - Total Reward: {total_reward:.2f} - Epsilon: {agent.epsilon:.2f}")

    print("\nTraining Complete.")
    
    # Test Run
    print("\nRunning Verification Episode (Greedy policy)...")
    state = env.reset()
    state[0] = 2.0 # Force start at 2.0
    state[1] = 0.0
    env.state = state # HACK: force env state
    
    print(f"Start State: {state}")
    for t in range(20):
        action = agent.policy(state) # Should likely be 0 (Push Left) if state is positive
        next_state, r, d = env.step(action)
        action_str = ["LEFT", "NONE", "RIGHT"][action]
        print(f"Time {t}: Pos={state[0]:.2f}, Vel={state[1]:.2f} -> Action: {action_str}")
        state = next_state
        if abs(state[0]) < 0.1 and abs(state[1]) < 0.1:
            print("Converged to Goal!")
            break

if __name__ == "__main__":
    run_dqn()

File: src_dev/reinforcement_learning/rl_dqn.jl
using Random
using Statistics
using Printf
using Flux

# Continuous 1D particle: state = (position, velocity); actions: left, none, right
mutable struct ContinuousWorld
    state::Vector{Float64}
    step::Int
    max_steps::Int
end

ContinuousWorld() = ContinuousWorld([0.0, 0.0], 0, 200)

actions() = (:left, :none, :right)

function reset!(env::ContinuousWorld)
    env.state = [rand(Random.default_rng()) * 4 - 2, 0.0]
    env.step = 0
    return env.state
end

function step!(env::ContinuousWorld, action::Int)
    pos, vel = env.state
    force = action == 1 ? -1.0 : action == 3 ? 1.0 : 0.0

    vel += 0.1 * force
    pos += 0.1 * vel

    pos = clamp(pos, -5.0, 5.0)
    vel = clamp(vel, -2.0, 2.0)

    env.state = [pos, vel]
    env.step += 1

    reward = -abs(pos) - 0.1 * abs(vel)
    done = env.step >= env.max_steps
    return env.state, reward, done
end

# Simple replay buffer implemented with a bounded vector
mutable struct ReplayBuffer
    data::Vector{Tuple{Vector{Float64},Int,Float64,Vector{Float64},Bool}}
    capacity::Int
end

ReplayBuffer(capacity::Int=10_000) = ReplayBuffer(Tuple{Vector{Float64},Int,Float64,Vector{Float64},Bool}[], capacity)

function push!(buf::ReplayBuffer, transition)
    push!(buf.data, transition)
    if length(buf.data) > buf.capacity
        popfirst!(buf.data)
    end
end

function sample(buf::ReplayBuffer, batch::Int, rng::AbstractRNG)
    idx = rand(rng, 1:length(buf.data), batch)
    return buf.data[idx]
end

# DQN agent with a tiny MLP
mutable struct DQNAgent
    net::Chain
    optim::Flux.Optimiser
    γ::Float64
    ε::Float64
    ε_decay::Float64
    ε_min::Float64
    batch::Int
    buffer::ReplayBuffer
end

function DQNAgent(state_size::Int, action_size::Int; γ=0.99, ε=1.0, ε_decay=0.995, ε_min=0.05,
                  batch=64, lr=1e-3, buffer_cap=10_000)
    net = Chain(Dense(state_size, 64, relu), Dense(64, 64, relu), Dense(64, action_size))
    optim = Flux.setup(Adam(lr), net)
    return DQNAgent(net, optim, γ, ε, ε_decay, ε_min, batch, ReplayBuffer(buffer_cap))
end

# Choose action with epsilon–greedy policy; actions are 1..3
function choose_action(agent::DQNAgent, state::Vector{Float64}, rng::AbstractRNG)
    if rand(rng) < agent.ε
        return rand(rng, 1:3)
    else
        q = agent.net(Flux.f32.(state))
        return argmax(q)
    end
end

function learn!(agent::DQNAgent, rng::AbstractRNG)
    buf = agent.buffer
    if length(buf.data) < agent.batch
        return
    end

    batch = sample(buf, agent.batch, rng)
    states  = Flux.f32.(hcat([t[1] for t in batch]...))  # 2 x B
    actions = [t[2] for t in batch]
    rewards = [t[3] for t in batch]
    nexts   = Flux.f32.(hcat([t[4] for t in batch]...))
    dones   = [t[5] for t in batch]

    function loss_fn()
        q_values = agent.net(states)                    # (3, B)
        chosen = Float32.([q_values[a, i] for (i, a) in enumerate(actions)])
        next_q = maximum(agent.net(nexts), dims=1)      # (1, B)
        targets = Float32.([rewards[i] + (dones[i] ? 0.0 : agent.γ * next_q[1, i]) for i in eachindex(rewards)])
        return Flux.Losses.mse(chosen, targets)
    end

    grads = Flux.gradient(loss_fn, Flux.params(agent.net))
    Flux.Optimise.update!(agent.optim, Flux.params(agent.net), grads)
end

function decay!(agent::DQNAgent)
    agent.ε = max(agent.ε_min, agent.ε * agent.ε_decay)
end

function run_dqn(; episodes=300, rng=Random.default_rng())
    env = ContinuousWorld()
    agent = DQNAgent(2, 3)

    println("Training DQN (Flux) to center the particle at zero...")
    for ep in 1:episodes
        state = reset!(env)
        total = 0.0
        done = false

        while !done
            action = choose_action(agent, state, rng)
            next_state, reward, done = step!(env, action)
            push!(agent.buffer, (state, action, reward, next_state, done))
            learn!(agent, rng)
            state = next_state
            total += reward
        end

        decay!(agent)
        if ep % 20 == 0
            @printf "Episode %4d/%4d  Reward: %7.2f  epsilon: %.3f\n" ep episodes total agent.ε
        end
    end

    println("\nGreedy rollout starting at position = 2.0")
    env.state = [2.0, 0.0]
    env.step = 0
    for t in 0:20
        pos, vel = env.state
        action = choose_action(agent, env.state, rng)
        name = (action == 1 ? "LEFT" : action == 2 ? "NONE" : "RIGHT")
        @printf "t=%2d  pos=%5.2f  vel=%5.2f  action=%s\n" t pos vel name
        _, _, done = step!(env, action)
        done && break
    end
end

if abspath(PROGRAM_FILE) == @__FILE__
    run_dqn()
end
